### ISSUE TITLE
New Term Request: Zstd binary compression

### ISSUE THREAD

**Issue by lazear** [HUPO-PSI/psi-ms-CV#167]
### Describe the new term or terms you would like to add.

Hi,  
I'd like to request a term for Zstd compression, analogous to zlib, but it can achieve greater performance.
http://facebook.github.io/zstd/

```
[Term]
id: MS:1000XXX
name: Zstd compression
def: "Zstd." [PSI:MS]
is_a: MS:1000572 ! binary data compression type
```

**Comment by mobiusklein** [comment]
For use in mzML, correct?

Ideally, this would be supported in ProteoWizard as well, given that it is a hub for file conversion, and most existing tools won't know what to do with this. It doesn't mean we can't add it though.

Do you plan to use this in an mzML converter with the ability to encode/decode to regular zlib compression as well? I looked at doing this a year or two ago but found limited benefit due to loss of portability.

**Comment by edeutsch** [comment]
I'm wondering if there's any real-world mzML cost-benefit analysis on this? Based on the table at the web site above, it would seem like the compression ratio gain is negligible. The compression/decompression speed definitely seems faster, but does it matter? is zlib fast enough that there's little practical benefit with ordinary reading and writing of mzMLs?

Are the benefits sufficiently high to overcome the cost of trying to get zstd into implementations? How does it compare with mzMLb, which we are poised to recommend?

I don't mean to put up road blocks, but wonder if there are already answers to these questions?

**Comment by lazear** [comment]
Hi Eric,

I was playing around with the compomics ThermoRawFileParser and got ~13% reduction in mzMl file size as well as a speed up by changing ~3 lines of code. There are some portability issues, but I think in general it should be pretty easy to drop in a zstd library.

I can do an analysis on a larger set of files and report back. 

Mike

**Comment by mobiusklein** [comment]
Generating some synthetic profile mass spectra, I have 12k data points for a single top-down isotopic pattern, which translates to about 0.1 MB of double precision floats. All caveats apply as I am testing this in Python.

Here's a microbenchmark of the stock `zlib` in the Python standard library, `isal_zlib`, a fork of Zlib with several "lessons learned" added from Intel (similar to `zlib-ng`), and `pyzstd`, a Python binding for the `zstd` C library:

## Compression
```python
In [33]: %timeit zlib.compress(mz_bytes)
5.49 ms ± 32 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [34]: %timeit isal_zlib.compress(mz_bytes)
193 µs ± 921 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [35]: %timeit pyzstd.compress(mz_bytes)
261 µs ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [43]: %timeit zlib.compress(mz_bytes, 9)
31.3 ms ± 117 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [44]: %timeit pyzstd.compress(mz_bytes, 9)
1.97 ms ± 3.88 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Compression Ratio (compressed / uncompressed)
| zlib | zlib-9 | isal_zlib | zstd | zstd-9 |
|-----| -------|----------|------|--------|
| 33%| 33%  |  52%      |  78%|  43% |

## Decompression

```python
In [60]: %timeit zlib.decompress(zlib_bytes)
221 µs ± 1.47 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [61]: %timeit zlib.decompress(zlib9_bytes)
220 µs ± 518 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [63]: %timeit isal_zlib.decompress(isal_bytes)
159 µs ± 447 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [64]: %timeit pyzstd.decompress(zstd_bytes)
88 µs ± 166 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [65]: %timeit pyzstd.decompress(zstd9_bytes)
180 µs ± 170 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Now, all of this is sensitive to the data you're compressing and how it is arranged. If I apply this to the intensity data which is much sparser (lots more zeros), things are more in `zstd`'s favor:

```python
In [69]: %timeit zlib.compress(intensity_bytes)
1.92 ms ± 2.62 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [70]: %timeit zlib.compress(intensity_bytes, 9)
1.94 ms ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [71]: %timeit isal_zlib.compress(intensity_bytes)
108 µs ± 319 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [72]: %timeit pyzstd.compress(intensity_bytes)
266 µs ± 1.49 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [73]: %timeit pyzstd.compress(intensity_bytes, 9)
606 µs ± 3.71 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Compression Ratio (compressed / uncompressed)
| zlib | zlib-9 | isal_zlib | zstd | zstd-9 |
|-----| -------|----------|------|--------|
| 95%| 95%  |  96%      |  96%|  96% |

## Inference
`zstd` gives much better performance than the stock `zlib` implementation at all levels of compression trading off slightly worse compression ratios (disk space being cheaper than they were 30 years ago, this is quite acceptable). However, the Intel `zlib` fork is almost as good while being inter-compatible with stock `zlib` (can mutually read what the other compresses and decompresses).

**Comment by mobiusklein** [comment]
RE Eric's last point about mzMLb, mzMLb already supports `zstd` compression transparently through HDF5, provided the filter is registered with the library at either build time or run time. Additionally `blosc`+`zstd` together produce the fastest compressed random access speed to space ratios around during testing on full raw files. 

Of course, mzMLb is still not widely implemented, though I'd expect that to change quickly if the ProteoWizard fork is merged.

**Comment by edeutsch** [comment]
Thanks!

I am struck by how much better the isal_zlib is that plain zlib. Seems like we should definitely be using that instead!

And also struck how little difference -9 makes (do I understand right that the "9" means the best compression setting?) Which is certainly slower. But no apparent size benefit?

**Comment by mobiusklein** [comment]
`-9` does tell `zlib` to work extra-hard to find longer pattern, it just couldn't find any. Maybe if I added some noise.

**Comment by lazear** [comment]
I processed 66 raw files using ThermoRawFileParser, using the latest 1.4.1 release, or my fork with default Zstd compression:

~14% reduction in file size, and ~30% reduction in run time. 
 
![image](https://user-images.githubusercontent.com/14057244/200932886-e9cc31eb-72d5-4f83-98a3-c43caa7daaca.png)

**Comment by lazear** [comment]
Indeed, that looks good. Feel free to close then, if there's no interest!

**Comment by mobiusklein** [comment]
I'm fine adding the new term, and for archival purposes people might choose to double the write time (remember I'm not showing the decompression speed here which I know `zstd` is better at. Does C# have `zlib-ng` bindings though? I know Rust does.

**Comment by lazear** [comment]
I'm not sure... I have enough C# knowledge to be dangerous to myself, and that's it. I would assume if it's possible to drop in a dynamically linked DLL (e.g. ISA-L or zlib-ng) it should be possible.

**Comment by mobiusklein** [comment]
https://github.com/carlst99/ZlibNGSharpMinimal

I've updated the PR with the Numpress cross-product.

**Comment by chambm** [comment]
> I'm fine adding the new term, and for archival purposes people might choose to double the write time (remember I'm not showing the decompression speed here which I know `zstd` is better at. Does C# have `zlib-ng` bindings though? I know Rust does.

Read speed improvement could be significant. How much faster are we talking for zstd?

**Comment by lazear** [comment]
I'll report back w/ read speeds for the 66 mzMLs (gzip/zstd) I produced yesterday. I can easily drop zlib-ng in, probably not isal_zlib though.

**Comment by lazear** [comment]
Zstd decreased total mzML parse times by 45% on average, compared to zlib-ng which decreased total parse times by 15%.

| Compression | Total time to read 66 mzMLs |
| -------------- | ------------------------------- |
| zlib | 828.7 s |
| zlib-ng | 680.1 s |
| zstd | 445.5 s |

![image](https://user-images.githubusercontent.com/14057244/201188843-3b70ace5-0ea7-487b-89cb-3fd8a6f252da.png)

**Comment by chambm** [comment]
That's very significant. But what about isal_zlib? I mean if it's the same or nearly the same speed for decompressing then there's not much reason to add another codec. @mobiusklein 's message makes it seem like it's not the same, but I don't understand why since ISAL is supposed to have acceleration for both compression and decompression.

**Comment by lazear** [comment]
It's gonna take me a bit to figure out how to test ISAL. zlib-ng is just changing a feature flag on the Rust package I'm using.

**Comment by lazear** [comment]
ISA-L does not export a zlib compatible API (outside of the python bindings) - there is no good pre-existing set of Rust bindings for ISA-L, so I set out to profile using pymzML instead. It appears to be **way** more work to use ISA-L than Zstd for non-python based implementations - also unclear how well it will perform on non-x86 based platforms (I won't be writing a set of bindings to test it out).

In pymzML, there is not as much of a substantial difference between the compression codecs (not sure what's going on there.... perhaps compression is no longer the rate-limiting step for a python-based parser?). I replaced all instances of  `zlib.decompress` with either `isal_zlib.decompress` or `pyzstd.decompress`
![image](https://user-images.githubusercontent.com/14057244/201250145-21cf6103-5292-4e90-94c3-284c7dc94109.png)

**Comment by mobiusklein** [comment]
Okay, once again, please remember we're testing with different languages, different files, and probably different "defaults", so performance ratios may not reflect across benchmarks.

In the interest of reducing implementation details driving divergence of our benchmarks, I am using https://ftp.pride.ebi.ac.uk/pride/data/archive/2016/04/PXD003498/AGP-tryp_O16_1.raw, which I am converting directly from vendor RAW format to mzML with a Python script using `ms_deisotope` as my conversion tool with `zstd` support hacked in for the evening, and then reading the mzML file back in using another short Python script to exercise read timing. I am not doing any peak picking or other data processing, just a straight conversion to mzML so I am storing each profile spectrum just as Thermo gives it to me.

Reading those profile mode mzML files I wrote last night:

| Compression | File Size | Time | Relative to `zlib` |
|----------------|----------|------| ------------------|
|No compression|4.47 GB | 40.28s| |
|`zlib` | 2.99 GB| 41.92s| |
|`isal_zlib` | 3.51 GB  | 41.25s |  98.4%|
|`zstd`|3.51GB|38.23s| 91.2% |
|`zstd-22`|2.75 GB|35.28s| 84.1% |
|`isal_zlib`-on-`zlib`| 2.99 GB | 38.93s|  92.8%  |

So by paying _double_ on write, `zstd` saved me 20% on read time on my local SSD while delivering a 10% reduction in size over `zlib`. Additionally, `zstd` still beat out `isal_zlib` on reading its sub-standard compression, but `isal_zlib` can read `zlib` trying really hard to compress the same data in about the same time `zstd` wrote it on it's "default" settings.

These results are likely a combination of both decompressor efficiency and the fact that I am reading half a gigabyte less of data off my local disk. 

As to the topic of `isal_zlib` vs `zlib-ng`, I suspect you're better off benchmarking writing with `zlib-ng` on your files than trying to get `isa-l` Rust bindings. Also, hopefully this reading benchmark drives home the benefits a bit more.

**Comment by mobiusklein** [comment]
mzML is just a poor vessel for a good compression algorithm that builds a large dictionary. Using mzMLb for the same file with blosc:zstd compression, I bring the file size down to 1.86 GB and can read the thing in 24.46 seconds. 

|Format| Size| Read Time|
|--------|-----|------------|
|Thermo RAW| 1.18 GB | 51.26s |
|mzML zlib | 2.99 GB | 41.92s |
|mzML zstd| 2.75 GB| 35.28s |
|mzMLb `blosc:zstd` | 1.86 GB | 24.46s |

So, rather than re-tooling every mzML parser to support `zstd` compressors, we really need to pool our time to get the mzMLb fork merged into ProteoWizard.

I know Rust's [HDF5 library](https://github.com/aldanor/hdf5-rust) has working `blosc` support, and when I wrote my own Rust mzML read/write library, I was just waiting for that to be released before starting to port my Python mzMLb writer there.
